### PR TITLE
TW-858: Jump to first unread msg when click on noti

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket:
- #858 

#### MOBILE

https://github.com/linagora/twake-on-matrix/assets/99852347/9e2e1c77-a06e-4fc8-bf01-cbf2f758d4e5

https://github.com/linagora/twake-on-matrix/assets/99852347/8ddf7c08-10ac-489d-8478-860c93f348d9

#### WEB

https://github.com/linagora/twake-on-matrix/assets/99852347/8634697c-8cd6-43ce-826b-f55bcc9998e2

https://github.com/linagora/twake-on-matrix/assets/99852347/a645036a-fa6a-4894-b94b-a9c68a1642ee

### CODE REVIEW

File: docs/adr/0015-callback-for-when-messages-become-visible-or-scroll-out-of-view.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# 14. Add a callback for when messages become visible or scroll out of view
```
Comment (hoangdat) [inline docs/adr/0015-callback-for-when-messages-become-visible-or-scroll-out-of-view.md:1]:
```suggestion
# 15. Add a callback for when messages become visible or scroll out of view
```

**Comment by sherlockvn** [comment]
What will happen if we only `setReadMarker` for the latest message? Because The SDK say that it updates the location, so i think the messages before the latest messages is mark read also ?
Screenshot 2023-12-14 at 01 47 58

File: lib/pages/chat/chat.dart [hunk 2]
```diff
@@ -331,19 +352,93 @@ class ChatController extends State<Chat>
     try {
       await loadTimelineFuture;
       final fullyRead = room?.fullyRead;
-      if (fullyRead == null || fullyRead == '') return;
-      if (timeline!.events.any((event) => event.eventId == fullyRead)) {
-        Logs().v('Scroll up to visible event', fullyRead);
+      if (fullyRead == null || fullyRead.isEmpty) {
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:355]:
```suggestion
      if (fullyRead == null || fullyRead.isEmpty || fullyRead == '') {
```

File: lib/pages/chat/chat.dart [hunk 3]
```diff
@@ -331,19 +352,93 @@ class ChatController extends State<Chat>
...
         setReadMarker();
         return;
       }
+      _firstTimeReadMessage = true;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:359]: `_visibilityDetectorAvailable`

File: lib/pages/chat/chat.dart [hunk 4]
```diff
@@ -331,19 +352,93 @@ class ChatController extends State<Chat>
...
+        lastIndexReadEvent != null) {
+      final afterFullyRead = timeline?.events.getRange(0, lastIndexReadEvent);
+      final unreadEvents = afterFullyRead
+          ?.where((event) => event.senderId != client.userID)
+          .toList();
+      if (unreadEvents == null || unreadEvents.isEmpty) return [];
+      Logs().d(
+        "Chat::getFirstUnreadEvent(): Last unread event ${unreadEvents.last}",
+      );
+      return unreadEvents;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:392]: no need this code?

File: lib/pages/chat/chat.dart [hunk 5]
```diff
@@ -187,6 +199,10 @@ class ChatController extends State<Chat>
           .firstWhereOrNull((eventId) => eventId == event.eventId) !=
       null;
 
+  double _currentVisibleFraction = 0;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:202]: no need

File: lib/pages/chat/chat.dart [hunk 6]
```diff
@@ -93,8 +93,18 @@ class ChatController extends State<Chat>
...
+
+  final _unreadMessageDebouncer = Debouncer(
+    _debouncerDuration,
+    initialValue: '',
+    checkEquality: false,
+  );
+
   static const double _isPortionAvailableToScroll = 64;
 
+  static const visibleFractionAvailable = 0.1;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:106]: remove. To check visibility, compare to 0 is enough

File: docs/adr/0014-find-to-last-read-message.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+Accepted
+
+## Context
+
+In the room, we have a lot of messages. When we open the room, we need to find the last read
+message.
+
+## Decision
+
+The user’s fully read marker is kept as an event in the room’s account data.
```
Comment (hoangdat) [inline docs/adr/0014-find-to-last-read-message.md:16]: It is different for each user in group.

File: lib/pages/chat/chat.dart [hunk 8]
```diff
@@ -109,6 +108,12 @@ class ChatController extends State<Chat>
 
   MatrixState? matrix;
 
+  Timer? _setReadMarkerTimer;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:111]: remove

File: lib/pages/chat/chat.dart [hunk 9]
```diff
@@ -187,6 +194,29 @@ class ChatController extends State<Chat>
           .firstWhereOrNull((eventId) => eventId == event.eventId) !=
       null;
 
+  String? _getMarkedUnreadLocationFilterOwn() {
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:197]:
```suggestion
  String? _findUnreadReceivedMessageLocation() {
```

File: lib/pages/chat/chat.dart [hunk 10]
```diff
@@ -109,6 +108,12 @@ class ChatController extends State<Chat>
 
   MatrixState? matrix;
 
+  Timer? _setReadMarkerTimer;
+
+  String _markerReadLocation = '';
+
+  String? makerReadLocationFiltered;
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:115]:
```suggestion
  String? _unreadReceivedMesssageLocation;
```

File: lib/pages/chat/chat.dart [hunk 11]
```diff
@@ -331,15 +360,16 @@ class ChatController extends State<Chat>
...
       final fullyRead = room?.fullyRead;
-      if (fullyRead == null || fullyRead == '') return;
-      if (timeline!.events.any((event) => event.eventId == fullyRead)) {
-        Logs().v('Scroll up to visible event', fullyRead);
+      if (fullyRead == null || fullyRead.isEmpty || fullyRead == '') {
         setReadMarker();
         return;
       }
+      _markerReadLocation = fullyRead;
+      makerReadLocationFiltered = _getMarkedUnreadLocationFilterOwn();
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:368]: Add to method `initUnreadLocation()`

File: lib/pages/chat/chat_view.dart [hunk 12]
```diff
@@ -127,7 +127,7 @@ class ChatView extends StatelessWidget with MessageContentMixin {
     }
 
     return GestureDetector(
-      onTapDown: controller.setReadMarker,
+      onTapDown: (_) => controller.setReadMarker(),
```
Comment (hoangdat) [inline lib/pages/chat/chat_view.dart:130]: please create ticket to improve `setReadMarker`

Comment (hoangdat) [reply lib/pages/chat/chat_view.dart:130]: dont' forget it

**Comment by hoangdat** [comment]
please rebase

File: lib/pages/chat/chat_view.dart (hunk 12 above)
Comment (nqhhdev) [reply lib/pages/chat/chat_view.dart:130]: Done #1147

**Review by hoangdat — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**